### ISSUE TITLE
sage: wrong hash for cypari2 patch

### ISSUE THREAD

**Issue by dylex** [NixOS/nixpkgs#55780]
## Issue description

The hash for `cypari2-2.0.3.patch` in pkgs/applications/science/math/sage/sage-src.nix appears to have changed.  (I believe #54690 would also fix this.)
```
hash mismatch in fixed-output derivation '/mnt/home/sw/nix/store/3g4ichm7slah3yy4a0rnzhk8p8v6r80s-cypari2-2.0.3.patch':
  wanted: sha256:08l2b9w0rn1zrha6188j72f7737xs126gkgmydjd31baa6367np2
  got:    sha256:0h9aj741db8n3njsp3rpmy4370gr8b2jzkmha5j6cv7gpcznf9fs
```

### Steps to reproduce

nix-build -A sage

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 3.10.0-957.1.3.el7.x86_64, CentOS Linux, 7 (Core)`
 - multi-user?: `no`
 - sandbox: `yes` (really no)
 - version: `nix-env (Nix) 2.2`

**Comment by dylex** [comment]
It does build and work fine with the updated hash.

**Comment by FRidh** [comment]
cc @timokau

**Comment by timokau** [comment]
That is surprising for multiple reasons

- I did specify the exact hash. Isn't `cgit`s `filterdiff` stable?
- Don't we mirror fixed-output derivations?

For reference (haven't looked into the differences yet), [here](https://gist.github.com/timokau/f4b3313b3c06bc67868e5f68b6c9f42f#file-cypari2-2-0-3-08l2b9w0rn1zrha6188j72f7737xs126gkgmydjd31baa6367np2-patch) is the old patch file from my nix store and [here](https://gist.github.com/timokau/f4b3313b3c06bc67868e5f68b6c9f42f#file-cypari2-2-0-3-08l2b9w0rn1zrha6188j72f7737xs126gkgmydjd31baa6367np2-patch) is the new one.

**Comment by timokau** [comment]
Looks like this is because I changed `fetchSageDiff` to use `cgit`s `rawdiff` instead of `patch`. No idea what I was thinking to do that without re-checking the hashes.

**Comment by dylex** [comment]
Those seem to be the same link.  I believe the file it's actually grabbing is https://git.sagemath.org/sage.git/rawdiff?id2=8.6.rc1&id=cd62d45bcef93fb4f7ed62609a46135e6de07051
This still appears to have some real (but probably unimportant) changes:
```
@@ -1,23 +1,3 @@
---- a/build/pkgs/cypari/checksums.ini
-+++ b/build/pkgs/cypari/checksums.ini
-@@ -1,4 +1,4 @@
- tarball=cypari2-VERSION.tar.gz
--sha1=26820d83ecb73555ddfb41b78ee87a3baff0bfbc
--md5=32f123679b6fc47c6713be2d9e0d2407
--cksum=4086011827
-+sha1=d25249b03ff0a0a2c15db0a33c94107b63c8a330
-+md5=8bfa43eea70fe6164c5be8453ff9f5ad
-+cksum=1869595633
---- a/build/pkgs/cypari/package-version.txt
-+++ b/build/pkgs/cypari/package-version.txt
-@@ -1 +1 @@
--1.3.1
-+2.0.3
---- a/build/pkgs/cypari/spkg-check
-+++ b/build/pkgs/cypari/spkg-check
-@@ -1 +1 @@
--cd src && $MAKE check
-+cd src && $MAKE check PYTHON=sage-python23
 --- a/src/sage/matrix/matrix_integer_dense.pxd
 +++ b/src/sage/matrix/matrix_integer_dense.pxd
 @@ -30,6 +30,5 @@
```

I also saw that the new file is incorrectly and randomly truncated, which seems to be some kind of caching problem with their server -- subtly altering the request temporarily fixes it, and I believe the `0h9aj...` hash is for the untruncated file.

**Comment by timokau** [comment]
Yes, I think the truncation is because they [use](https://groups.google.com/forum/#!topic/sage-devel/SoIBiKbeLFM) an outdated cgit and `filterdiff` is broken. The other difference is because of the changed `exclude`. Should be fixed in #55783. Thanks for the report.
